crates/relayer-solomachine/src/impls/chain.rs

```diff
@@ -0,0 +1,1365 @@
+            new_diversifier: next_diversifier,
+        };
+
+        let sign_data = SolomachineSignHeaderData {
+            header_data,
+            sequence: client_state.sequence,
+            timestamp,
+            diversifier: current_diversifier.clone(),
+        };
+
+        let secret_key = self.chain.secret_key();
+
+        let signature = sign_header_data(secret_key, &sign_data).map_err(Chain::encode_error)?;
+
+        let header = SolomachineHeader {
+            timestamp,
+            signature,
+            header_data: sign_data.header_data,
+        };
+
+        let payload = SolomachineUpdateClientPayload { header };
+
+        Ok(payload)
+    }
+
+    async fn build_connection_open_init_payload(
+        &self,
+        _client_state: &SolomachineClientState,
+    ) -> Result<SolomachineConnectionOpenInitPayload, Chain::Error> {
+        let commitment_prefix = self.chain.commitment_prefix();
+
+        let payload = SolomachineConnectionOpenInitPayload {
+            commitment_prefix: commitment_prefix.to_string(),
+        };
+
+        Ok(payload)
+    }
+
+    async fn build_connection_open_try_payload(
+        &self,
+        solo_client_state: &SolomachineClientState,
+        height: &Height,
+        client_id: &ClientId,
+        connection_id: &ConnectionId,
+    ) -> Result<SolomachineConnectionOpenTryPayload, Chain::Error> {
+        let connection = self.chain.query_connection(connection_id).await?;
+
+        if connection.state != ConnectionState::Init {
+            return Err(Chain::invalid_connection_state_error(
+                ConnectionState::Init,
+                connection.state,
+            ));
+        }
+
+        let versions = connection.versions().to_vec();
+
+        let delay_period = connection.delay_period();
+
+        let commitment_prefix = self.chain.commitment_prefix();
+
+        let connection_data = connection_proof_data(
+            solo_client_state,
+            commitment_prefix,
+            connection_id,
+            connection,
+        )
+        .map_err(Chain::encode_error)?;
+
+        let cosmos_client_state = self.chain.query_client_state(client_id).await?;
+
+        let client_state_data = client_state_proof_data(
+            solo_client_state,
+            commitment_prefix,
+            client_id,
+            &cosmos_client_state,
+        )
+        .map_err(Chain::encode_error)?;
+
+        let cosmos_consensus_state = self.chain.query_consensus_state(client_id, *height).await?;
+
+        let consensus_state_data = consensus_state_proof_data(
+            solo_client_state,
+            commitment_prefix,
+            client_id,
+            *height,
+            &cosmos_consensus_state,
+        )
+        .map_err(Chain::encode_error)?;
+
+        let secret_key = self.chain.secret_key();
+
+        let connection_proof =
+            sign_with_data(secret_key, &client_state_data).map_err(Chain::encode_error)?;
```

Comment:
```suggestion
            sign_with_data(secret_key, &connection_data).map_err(Chain::encode_error)?;
```
Shouldn't this be the proof for the connection_data?

Reply:
Probably.

Reply:
[7737304](https://github.com/informalsystems/hermes/pull/3509/commits/7737304fc38b281d5548de134c5b8e525400a928)